cpp/src/io/utilities/column_utils.cuh

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+namespace cudf {
+namespace io {
+
+/**
+ * @brief Create column_device_view pointers from leaf columns
+ *
+ * A device_uvector is created to store the leaves of parent columns. The
+ * column descriptor array is updated to point to these leaf columns.
+ *
+ * @param col_desc Column description array
+ * @param parent_table_device_view Table device view containing parent columns
+ * @param stream CUDA stream to use
+ *
+ * @return Device array containing leaf column device views
+ */
+template <typename ColumnDescriptor>
+rmm::device_uvector<column_device_view> create_leaf_column_device_views(
+  hostdevice_vector<ColumnDescriptor> &col_desc,
```

Comment:
> only device buffer is used

did you mean inside this function or before this function is called?

Should I explicitly convert the `hostdevice_vector` to `device_span` using
`constexpr span_base(pointer data, size_type size);`

or add support so that `hostdevice_vector` is implicitly converted to a `device_span`?

Reply:
Here's what I added to `hostdevice_vector` in [#7324](https://github.com/rapidsai/cudf/pull/7324/files?file-filters%5B%5D=.h&file-filters%5B%5D=.hpp#diff-a97b18b281b557fd56b44267cd57cda53be25f7073a6a86b34c083d88658ac64R95):
```
  operator cudf::detail::device_span<T>() { return {d_data.data(), max_elements}; }
  operator cudf::detail::device_span<T const>() const { return {d_data.data(), max_elements}; }

  operator cudf::detail::host_span<T>() { return {h_data, max_elements}; }
  operator cudf::detail::host_span<T const>() const { return {h_data, max_elements}; }
```
Feel free to make the same change. Otherwise, let me look into merging https://github.com/rapidsai/cudf/pull/7324 ASAP.